File: uni_modules/ui-component/utssdk/app-ios/swift/observer/DeviceStoreObserver.swift
import Combine
import DCloudUTSFoundation
import RTCRoomEngine_Plus

public class DeviceStoreObserver {
    private var cancellables = Set<AnyCancellable>()
    public static let shared = DeviceStoreObserver()
        
    public func deviceStoreChanged(_ callback: @escaping (_ name: String, _ data: String) -> Void) {
        cancellables.removeAll()

        V2RTCRoomEngine.shared.getDeviceStore()
            .state.subscribe(StateSelector(keyPath: \V2DeviceState.microphoneStatus))
            .removeDuplicates()
            .receive(on: DispatchQueue.main)
            .sink(receiveValue: { [weak self] value in
                guard let self = self else { return }
                callback("microphoneStatus", String(value.rawValue))
            })
            .store(in: &cancellables)

        V2RTCRoomEngine.shared.getDeviceStore()
            .state.subscribe(StateSelector(keyPath: \V2DeviceState.microphoneStatusReason))
            .removeDuplicates()
            .receive(on: DispatchQueue.main)
            .sink(receiveValue: { [weak self] value in
                guard let self = self else { return }
                callback("microphoneStatusReason", String(value.rawValue))
            })
            .store(in: &cancellables)

        V2RTCRoomEngine.shared.getDeviceStore()
            .state.subscribe(StateSelector(keyPath: \V2DeviceState.microphoneLastError))
            .removeDuplicates()
            .receive(on: DispatchQueue.main)
            .sink(receiveValue: { [weak self] value in
                guard let self = self else { return }
                callback("microphoneLastError", String(value.rawValue))
            })
            .store(in: &cancellables)

        V2RTCRoomEngine.shared.getDeviceStore()
            .state.subscribe(StateSelector(keyPath: \V2DeviceState.captureVolume))
            .removeDuplicates()
            .receive(on: DispatchQueue.main)
            .sink(receiveValue: { [weak self] value in
                guard let self = self else { return }
                callback("captureVolume", String(value))
            })
            .store(in: &cancellables)

        V2RTCRoomEngine.shared.getDeviceStore()
            .state.subscribe(StateSelector(keyPath: \V2DeviceState.cameraStatus))
            .removeDuplicates()
            .receive(on: DispatchQueue.main)
            .sink(receiveValue: { [weak self] value in
                guard let self = self else { return }
                callback("cameraStatus", String(value.rawValue))
            })
            .store(in: &cancellables)

        V2RTCRoomEngine.shared.getDeviceStore()
            .state.subscribe(StateSelector(keyPath: \V2DeviceState.cameraStatusReason))
            .removeDuplicates()
            .receive(on: DispatchQueue.main)
            .sink(receiveValue: { [weak self] value in
                guard let self = self else { return }
                callback("cameraStatusReason", String(value.rawValue))
            })
            .store(in: &cancellables)

        V2RTCRoomEngine.shared.getDeviceStore()
            .state.subscribe(StateSelector(keyPath: \V2DeviceState.isFrontCamera))
            .removeDuplicates()
            .receive(on: DispatchQueue.main)
            .sink(receiveValue: { [weak self] value in
                guard let self = self else { return }
                callback("isFrontCamera", String(value))
            })
            .store(in: &cancellables)

        V2RTCRoomEngine.shared.getDeviceStore()
            .state.subscribe(StateSelector(keyPath: \V2DeviceState.isLocalMirror))
            .removeDuplicates()
            .receive(on: DispatchQueue.main)
            .sink(receiveValue: { [weak self] value in
                guard let self = self else { return }
                callback("isLocalMirror", String(value))
            })
            .store(in: &cancellables)

        V2RTCRoomEngine.shared.getDeviceStore()
            .state.subscribe(StateSelector(keyPath: \V2DeviceState.localVideoQuality))
            .removeDuplicates()
            .receive(on: DispatchQueue.main)
            .sink(receiveValue: { [weak self] value in
                guard let self = self else { return }
                guard let value = value else { return }
                callback("localVideoQuality", convertVideoQuality(value))
            })
            .store(in: &cancellables)

        V2RTCRoomEngine.shared.getDeviceStore()
            .state.subscribe(StateSelector(keyPath: \V2DeviceState.cameraLastError))
            .removeDuplicates()
            .receive(on: DispatchQueue.main)
            .sink(receiveValue: { [weak self] value in
                guard let self = self else { return }
                callback("cameraLastError", String(value.rawValue))
            })
            .store(in: &cancellables)

        V2RTCRoomEngine.shared.getDeviceStore()
            .state.subscribe(StateSelector(keyPath: \V2DeviceState.currentAudioRoute))
            .removeDuplicates()
            .receive(on: DispatchQueue.main)
            .sink(receiveValue: { [weak self] value in
                guard let self = self else { return }
                callback("currentAudioRoute", String(value.rawValue))
            })
            .store(in: &cancellables)

        V2RTCRoomEngine.shared.getDeviceStore()
            .state.subscribe(StateSelector(keyPath: \V2DeviceState.screenStatus))
            .removeDuplicates()
            .receive(on: DispatchQueue.main)
            .sink(receiveValue: { [weak self] value in
                guard let self = self else { return }
                callback("screenStatus", String(value.rawValue))
            })
            .store(in: &cancellables)

        V2RTCRoomEngine.shared.getDeviceStore()
            .state.subscribe(StateSelector(keyPath: \V2DeviceState.screenStatusReason))
            .removeDuplicates()
            .receive(on: DispatchQueue.main)
            .sink(receiveValue: { [weak self] value in
                guard let self = self else { return }
                callback("screenStatusReason", String(value.rawValue))
            })
            .store(in: &cancellables)

        V2RTCRoomEngine.shared.getDeviceStore()
            .state.subscribe(StateSelector(keyPath: \V2DeviceState.networkInfo))
            .removeDuplicates()
            .receive(on: DispatchQueue.main)
            .sink(receiveValue: { [weak self] value in
                guard let self = self else { return }
                if let json = JsonUtil.toJson(convertNetworkInfo(value)) {
                    callback("networkInfo", json)
                }
            })
            .store(in: &cancellables)
    }
    
    private func convertNetworkInfo(_ info: TUINetworkInfo) -> [String: Any] {
        var map = [String: Any]()
        map["userId"] = info.userId ?? ""
        map["quality"] = convertNetworkQuality(info.quality)
        map["upLoss"] = info.upLoss
        map["downLoss"] = info.downLoss
        map["delay"] = info.delay
        return map
    }
    
    private func convertNetworkQuality(_ quality: TUINetworkQuality?) -> String {
        guard let quality = quality else {
            return "UNKNOWN"
        }
        switch quality {
            case .veryBad:
                return "VBAD"
            case .bad:
                return "BAD"
            case .down:
                return "DOWN"
            case .poor:
                return "POOR"
            case .good:
                return "GOOD"
            case .excellent:
                return "EXCELLENT"
            default:
                return "UNKNOWN"
        }
    }
    
    private func convertVideoQuality(_ quality: TUIVideoQuality) -> String {
        switch quality {
            case TUIVideoQuality.quality540P:
                return "540P"
            case TUIVideoQuality.quality720P:
                return "720P"
            case TUIVideoQuality.quality1080P:
                return "1080P"
            default:
                return "360P"
        }
    }
} 

File: uni_modules/ui-component/utssdk/app-android/kotlin/observer/DeviceStoreObserver.kt
package uts.sdk.modules.uiComponent.observer

import com.google.gson.Gson
import com.tencent.cloud.tuikit.engine.common.TUICommonDefine
import com.tencent.cloud.tuikit.engine.room.TUIRoomDefine
import com.tencent.cloud.uikit.state.DeviceError
import com.tencent.cloud.uikit.state.DeviceStatus
import com.tencent.cloud.uikit.state.DeviceStatusReason
import com.tencent.cloud.uikit.state.V2RTCRoomEngine
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.Job
import kotlinx.coroutines.launch

object DeviceStoreObserver {
    private val gson = Gson()
    private var bindDataJob: Job? = null

    fun deviceStoreChanged(callback: (String, String) -> Unit) {
        bindDataJob?.cancel()
        bindDataJob = CoroutineScope(Dispatchers.Main).launch {
            launch {
                V2RTCRoomEngine.getInstance().getDeviceStore().deviceState.microphoneStatus.collect { status ->
                    val micStatus = if (status != null) status else DeviceStatus.OFF
                    callback("microphoneStatus", gson.toJson(micStatus.value))
                }
            }
            launch {
                V2RTCRoomEngine.getInstance().getDeviceStore().deviceState.microphoneStatusReason.collect { micReason ->
                    val reason = if (micReason != null) micReason else DeviceStatusReason.CHANGED_BY_SELF
                    callback("microphoneStatusReason", gson.toJson(reason.value))
                }
            }

            launch {
                V2RTCRoomEngine.getInstance().getDeviceStore().deviceState.microphoneLastError.collect { deviceError ->
                    val error = if (deviceError != null) deviceError else DeviceError.NO_ERROR
                    callback("microphoneLastError", gson.toJson(error.value))
                }
            }
            launch {
                V2RTCRoomEngine.getInstance().getDeviceStore().deviceState.currentAudioRoute.collect { audioRoute ->
                    callback("currentAudioRoute", gson.toJson(audioRoute.value))
                }
            }
            launch {
                V2RTCRoomEngine.getInstance().getDeviceStore().deviceState.captureVolume.collect { volume ->
                    callback("captureVolume", gson.toJson(volume))
                }
            }

            launch {
                V2RTCRoomEngine.getInstance().getDeviceStore().deviceState.cameraStatus.collect { cameraStatus ->
                    callback("cameraStatus", gson.toJson(cameraStatus.value))
                }
            }

            launch {
                V2RTCRoomEngine.getInstance().getDeviceStore().deviceState.cameraStatusReason.collect { reason ->
                    callback("cameraStatusReason", gson.toJson(reason.value))
                }
            }

            launch {
                V2RTCRoomEngine.getInstance().getDeviceStore().deviceState.cameraLastError.collect { deviceError ->
                    val error = if (deviceError != null) deviceError else DeviceError.NO_ERROR
                    callback("cameraLastError", gson.toJson(error.value))
                }
            }

            launch {
                V2RTCRoomEngine.getInstance().getDeviceStore().deviceState.isFrontCamera.collect { isFrontCamera ->
                    callback("isFrontCamera", gson.toJson(isFrontCamera))
                }
            }
            launch {
                V2RTCRoomEngine.getInstance().getDeviceStore().deviceState.isLocalMirror.collect { isLocalMirror ->
                    callback("isLocalMirror", gson.toJson(isLocalMirror))
                }
            }

            launch {
                V2RTCRoomEngine.getInstance().getDeviceStore().deviceState.localVideoQuality.collect { quality ->
                    callback("localVideoQuality", gson.toJson(convertVideoQuality(quality)))
                }
            }
            launch {
                V2RTCRoomEngine.getInstance().getDeviceStore().deviceState.screenStatus.collect { screenStatus ->
                    callback("screenStatus", gson.toJson(screenStatus.value))
                }
            }
            launch {
                V2RTCRoomEngine.getInstance().getDeviceStore().deviceState.screenStatus.collect { screenStatusReason ->
                    callback("screenStatusReason", gson.toJson(screenStatusReason.value))
                }
            }
            launch {
                V2RTCRoomEngine.getInstance().getDeviceStore().deviceState.networkInfo.collect { networkInfo ->
                    callback("networkInfo", gson.toJson(convertNetworkInfo(networkInfo)))
                }
            }
        }
    }

    private fun convertNetworkInfo(info: TUICommonDefine.NetworkInfo): Map<String, Any> {
        val map = mutableMapOf<String, Any>()
        map["userId"] = info.userId ?: ""
        map["quality"] = convertNetworkQuality(info.quality)
        map["upLoss"] = info.upLoss
        map["downLoss"] = info.downLoss
        map["delay"] = info.delay
        return map
    }

    private fun convertNetworkQuality(quality: TUICommonDefine.NetworkQuality?): String {
        if (quality == null) return "UNKNOWN"
        return when (quality) {
            TUICommonDefine.NetworkQuality.VERY_BAD -> "VBAD"
            TUICommonDefine.NetworkQuality.BAD -> "BAD"
            TUICommonDefine.NetworkQuality.DOWN -> "DOWN"
            TUICommonDefine.NetworkQuality.POOR -> "POOR"
            TUICommonDefine.NetworkQuality.GOOD -> "GOOD"
            TUICommonDefine.NetworkQuality.EXCELLENT -> "EXCELLENT"
            else ->
                "UNKNOWN"
        }
    }

    private fun convertVideoQuality(quality: TUIRoomDefine.VideoQuality): String {
        return when (quality) {
            TUIRoomDefine.VideoQuality.Q_540P -> "540P"
            TUIRoomDefine.VideoQuality.Q_720P -> "720P"
            TUIRoomDefine.VideoQuality.Q_1080P -> "1080P"
            else -> "360P"
        }
    }
}